deepinv/physics/mcmri.py

```diff
@@ -0,0 +1,77 @@
+import torch
+from torch import nn
+from deepinv.physics.forward import LinearPhysics
+
+from deepinv.physics import adjoint_function
+
+
+
+
+class MultiCoilMRI(LinearPhysics):
+    r"""
+    Multi-coil MRI operator.
+
+    The linear operator is defined as:
+    .. math::
+
+        y_c = \text{diag}(p) F \text{diag}(s_c) x
+
+        for c=1,\dots,C coils, where y_c are the measurements from the cth coil, \text{diag}(p) is the acceleration mask, F is the Fourier transform and \diag(s_c) is the cth coil sensitivity.
+    
+    :param torch.Tensor mask: binary sampling mask which should have shape [B,1,H,W].
+    :param torch.Tensor coil_maps: complex valued coil sensitvity maps which should have shape [B,C,H,W].
+    :param device: specify which device you want to use (i.e, cpu or gpu).
+    """
+
+    def __init__(
+        self,
+        mask,
+        coil_maps,
+        device=torch.device("cpu"),
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        self.mask = mask.to(device)
+        self.coil_maps = coil_maps.to(device)
+
+    def A(self, x, **kwargs):
+        r"""  
+        Applies linear operator.  
+        
+        :param torch.Tensor x: image with shape [B,H,W,2].  
+        :returns: (torch.Tensor) multi-coil kspace measurements with shape [B,C,H,W,2].  
```

Comment:
For 5D data I think the shape should either be `B,2,H,W,N` or `B,2,N,H,W`, where N is the coil dimension. I would prefer `B,2,N,H,W` as it keeps the spatial dimensions at the end which is consistent with code elsewhere (e.g. in `MRI.check_mask`, or wherever we take `H, W = x.shape[-2:]`). 

For example, in `DynamicMRI`, the 5D data is of shape `B,2,T,H,W`. This is readily compatible with `check_mask` (see `DynamicMRI.check_mask`. @tachella what do you think?

Reply:
I agree, let's go for `B,2,N,H,W`

Reply:
By using N do we intend to make this work for 3D multi-coil? If so I also need to change the forward and adjoint to account for this. As it currently stands this operator is assuming 2D multi-slice data

Reply:
I haven't tested the current single coil MRI with 3D volumes yet. I think if we want to add 3D volume support we should make sure single coil MRI (and possibly dynamic MRI) also support 3D volumes? And secondly, if so, this would make this 6D data, so we would decide whether we want `B,2,N,D,H,W` or `D` at the end, where `D` is the depth dimension.

Reply:
This has been sorted now so we can conditionally accept 2D or 3D using the `three_d` parameter